[WIP] Replacing port-forwarding with k8s events in recordevents

slinkydeveloper opened the pull request:
Signed-off-by: Francesco Guardiani <francescoguard@gmail.com>

Fixes #3570

<!-- Please include the 'why' behind your changes if no issue exists -->

## Proposed Changes

- This PR takes the ideas from pkg/test/observer and ports them to recordevents, without changing the API. This change is meant to be transparent both here in eventing and in downstream projects.

mattmoor commented on test/lib/recordevents/event_info.go:
```diff
@@ -62,11 +62,14 @@ type MinMaxResponse struct {
 	// HTTPHeaders of the connection that delivered the event
-	HTTPHeaders map[string][]string
+	HTTPHeaders map[string][]string `json:"httpHeaders"`
+	Origin   string            `json:"origin"`
+	Observer string            `json:"observer"`
+	Time     time.Time            `json:"time"`
```
<!--gotool-bot-->
Format Go code:
```suggestion
	Origin      string              `json:"origin"`
	Observer    string              `json:"observer"`
	Time        time.Time           `json:"time"`
```

mattmoor commented on test/lib/recordevents/observer/observer.go:
```diff
@@ -0,0 +1,85 @@
+// EventLog instances. It will listen on :8080.
+func New(name string, eventLogs ...recordevents.EventLog) *Observer {
+	return &Observer{
+		Name: name,
```
<!--gotool-bot-->
Format Go code:
```suggestion
		Name:      name,
```

mattmoor reviewed (commented):
<!--gotool-bot-->
Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -path './third_party' -prune -o -name '*.pb.go' -prune -o -type f -name '*.go' -print)`
  `goimports -w $(find -name '*.go' | grep -v vendor | grep -v third_party | grep -v .pb.go | grep -v wire_gen.go)`

mattmoor commented on test/conformance/helpers/channel_status_subscriber_test_helper.go:
```diff
@@ -59,8 +60,7 @@ func channelHasRequiredSubscriberStatus(ctx context.Context, st *testing.T, clie
 
-	pod := resources.EventRecordPod(subscriberServiceName + "-pod")
-	client.CreatePodOrFail(pod, testlib.WithService(subscriberServiceName))
+	_ = recordevents.DeployEventRecordOrFail(context.TODO(), client, subscriberServiceName + "-pod")
```
<!--gotool-bot-->
Format Go code:
```suggestion
	_ = recordevents.DeployEventRecordOrFail(context.TODO(), client, subscriberServiceName+"-pod")
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
	_ = recordevents.DeployEventRecordOrFail(context.TODO(), client, subscriberServiceName+"-pod")
```

mattmoor commented on test/lib/recordevents/event_info_store.go:
```diff
@@ -100,15 +100,35 @@ func NewEventInfoStore(client *testlib.Client, podName string) (*EventInfoStore,
+		resources.WithRuleForRole(&rbacv1.PolicyRule{
+			APIGroups: []string{""},
+			Resources: []string{"pods"},
+			Verbs: []string{"get", "list", "watch"},
```
<!--gotool-bot-->
Format Go code:
```suggestion
			Verbs:     []string{"get", "list", "watch"},
```

mattmoor commented on test/lib/recordevents/event_info_store.go:
```diff
@@ -100,15 +100,35 @@ func NewEventInfoStore(client *testlib.Client, podName string) (*EventInfoStore,
+		resources.WithRuleForRole(&rbacv1.PolicyRule{
+			APIGroups: []string{""},
+			Resources: []string{"events"},
+			Verbs: []string{rbacv1.VerbAll},
```
<!--gotool-bot-->
Format Go code:
```suggestion
			Verbs:     []string{rbacv1.VerbAll},
```

mattmoor commented on test/lib/resources/kube.go:
```diff
@@ -53,7 +53,18 @@ func eventLoggerPod(imageName string, name string) *corev1.Pod {
 				Image:           pkgTest.ImagePath(imageName),
 				ImagePullPolicy: corev1.PullAlways,
+				Env: []corev1.EnvVar{{
+					Name: "SYSTEM_NAMESPACE",
```
<!--gotool-bot-->
Format Go code:
```suggestion
					Name:  "SYSTEM_NAMESPACE",
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
					Name:  "SYSTEM_NAMESPACE",
```

mattmoor commented on test/lib/resources/kube.go:
```diff
@@ -53,7 +53,18 @@ func eventLoggerPod(imageName string, name string) *corev1.Pod {
+					Name: "SYSTEM_NAMESPACE",
+					Value: namespace,
+				}, {
+					Name: "OBSERVER",
```
<!--gotool-bot-->
Format Go code:
```suggestion
					Name:  "OBSERVER",
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
					Name:  "OBSERVER",
```

mattmoor commented on test/lib/resources/kube.go:
```diff
@@ -53,7 +53,18 @@ func eventLoggerPod(imageName string, name string) *corev1.Pod {
+				}},
 			}},
+			ServiceAccountName: serviceAccountName,
 			RestartPolicy: corev1.RestartPolicyAlways,
```
<!--gotool-bot-->
Format Go code:
```suggestion
			RestartPolicy:      corev1.RestartPolicyAlways,
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
			RestartPolicy:      corev1.RestartPolicyAlways,
```

mattmoor reviewed (commented):
<!--gotool-bot-->
Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -path './third_party' -prune -o -name '*.pb.go' -prune -o -type f -name '*.go' -print)`
  `goimports -w $(find -name '*.go' | grep -v vendor | grep -v third_party | grep -v .pb.go | grep -v wire_gen.go)`

mattmoor reviewed (commented):
<!--gotool-bot-->
Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -path './third_party' -prune -o -name '*.pb.go' -prune -o -type f -name '*.go' -print)`
  `goimports -w $(find -name '*.go' | grep -v vendor | grep -v third_party | grep -v .pb.go | grep -v wire_gen.go)`

mattmoor commented on test/conformance/helpers/channel_status_subscriber_test_helper.go:
```diff
@@ -59,8 +60,7 @@
 
-	pod := resources.EventRecordPod(subscriberServiceName + "-pod")
-	client.CreatePodOrFail(pod, testlib.WithService(subscriberServiceName))
+	_ = recordevents.DeployEventRecordOrFail(context.TODO(), client, subscriberServiceName + "-pod")
```
<!--gotool-bot-->
Format Go code:
```suggestion
	_ = recordevents.DeployEventRecordOrFail(context.TODO(), client, subscriberServiceName+"-pod")
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
	_ = recordevents.DeployEventRecordOrFail(context.TODO(), client, subscriberServiceName+"-pod")
```

mattmoor commented on test/lib/resources/kube.go:
```diff
@@ -53,7 +53,18 @@
 				Image:           pkgTest.ImagePath(imageName),
 				ImagePullPolicy: corev1.PullAlways,
+				Env: []corev1.EnvVar{{
+					Name: "SYSTEM_NAMESPACE",
```
<!--gotool-bot-->
Format Go code:
```suggestion
					Name:  "SYSTEM_NAMESPACE",
```

mattmoor commented on test/lib/resources/kube.go:
```diff
@@ -53,7 +53,18 @@
+					Name: "SYSTEM_NAMESPACE",
+					Value: namespace,
+				}, {
+					Name: "OBSERVER",
```
<!--gotool-bot-->
Format Go code:
```suggestion
					Name:  "OBSERVER",
```

mattmoor commented on test/lib/resources/kube.go:
```diff
@@ -53,7 +53,18 @@
+				}},
 			}},
+			ServiceAccountName: serviceAccountName,
 			RestartPolicy: corev1.RestartPolicyAlways,
```
<!--gotool-bot-->
Format Go code:
```suggestion
			RestartPolicy:      corev1.RestartPolicyAlways,
```

mattmoor reviewed (commented):
<!--gotool-bot-->
Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -path './third_party' -prune -o -name '*.pb.go' -prune -o -type f -name '*.go' -print)`
  `goimports -w $(find -name '*.go' | grep -v vendor | grep -v third_party | grep -v .pb.go | grep -v wire_gen.go)`

mattmoor commented on test/lib/recordevents/resources.go:
```diff
@@ -0,0 +1,43 @@
+import (
+	"fmt"
+
+	"k8s.io/api/core/v1"
```
<!--gotool-bot-->
Format Go code:
```suggestion
	v1 "k8s.io/api/core/v1"
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
	v1 "k8s.io/api/core/v1"
```

mattmoor reviewed (commented):
<!--gotool-bot-->
Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -path './third_party' -prune -o -name '*.pb.go' -prune -o -type f -name '*.go' -print)`
  `goimports -w $(find -name '*.go' | grep -v vendor | grep -v third_party | grep -v .pb.go | grep -v wire_gen.go)`

mattmoor reviewed (commented):
<!--gotool-bot-->
Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -path './third_party' -prune -o -name '*.pb.go' -prune -o -type f -name '*.go' -print)`
  `goimports -w $(find -name '*.go' | grep -v vendor | grep -v third_party | grep -v .pb.go | grep -v wire_gen.go)`

slinkydeveloper commented:
/assign @lionelvillard @n3wscott 

@n3wscott Can i remove the code under pkg/test/observer? At this point that code is orphan and not used in this repo (I've moved the bits i needed and modified them under test/lib/recordevents)

lionelvillard commented on test/lib/k8s_events.go:
```diff
@@ -0,0 +1,84 @@
+// NewEventListener creates a new event listener
+func NewEventListener(client kubernetes.Interface, namespace string, logf func(string, ...interface{})) *EventListener {
+	ctx, cancelCtx := context.WithCancel(context.Background())
+	informerFactory := informers.NewSharedInformerFactoryWithOptions(
+		client,
+		0,
+		informers.WithNamespace(namespace),
+	)
+	eventsInformer := informerFactory.Core().V1().Events().Informer()
+
+	el := EventListener{
+		cancel: cancelCtx,
+	}
+
+	eventsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			el.handle(obj.(*corev1.Event))
+		},
+	})
+
+	go func() {
+		eventsInformer.Run(ctx.Done())
+		logf("EventListener stopped")
+	}()
```
In reconciler-test the event informer is injected into the context (something to keep in mind when moving this code over).

slinkydeveloper replied:
Cool so we can avoid creating it manually!

slinkydeveloper replied:
We'll still need this `EventListener`, right?

lionelvillard replied:
Yup I think so.

n3wscott commented on test/lib/recordevents/event_log.go:
```diff
@@ -0,0 +1,22 @@
+package recordevents
+
+// EventLog is the contract for an event logger to vent an event.
+type EventLog interface {
```
why are you copying this to `test/lib`?

n3wscott replied:
now we have three copies of this code?

slinkydeveloper replied:
Because that code is hard to reuse, I had to do some important changes (interfaces and impls looks similar, but they're different)

slinkydeveloper replied:
That's why I'm saying, if we use that code only in recordevents, what if we just remove the one under `test/lib`?

slinkydeveloper replied:
> now we have three copies of this code?

Copying to the new test framework was a mistake, i'm going to port this new code there :smile:

slinkydeveloper commented:
The issue here is weird, for some reason it timeouts on waiting for pods to delete

slinkydeveloper commented:
Solved, I was not closing the http server correctly

slinkydeveloper commented:
/retest

n3wscott commented on test/lib/recordevents/recorder_vent/constructor.go:
```diff
@@ -72,7 +72,12 @@ func createRecorder(ctx context.Context, agentName string) record.EventRecorder
+			MessageFunc: func(event *corev1.Event) string {
+				return event.Message
+			},
+			MaxEvents: 1,
```
NICE FIND!!!

n3wscott reviewed (commented):
/lgtm
/approve

matzew commented:
@slinkydeveloper is it possible to backport this to 0.17 and 0.18 ? :grin:

slinkydeveloper commented:
@matzew Not sure, changes are too much extensive both on recordevents and lib test code... There are also other changes in that code before this one
